src/backend/distributed/planner/modify_planner.c

```diff
@@ -544,6 +568,86 @@ DistributedModifyShardInterval(Query *query)
 
 
 /*
+ * UseFastShardPruning returns true if the commandType is INSERT and partition method
+ * is hash or range.
+ */
+static bool
+UseFastShardPruning(CmdType commandType, char partitionMethod)
+{
+	/* we currently only support INSERTs */
+	if (commandType != CMD_INSERT)
+	{
+		return false;
+	}
+
+	/* INSERTs could only go into hash and range distributed tables */
+	if (partitionMethod == DISTRIBUTE_BY_HASH || partitionMethod == DISTRIBUTE_BY_RANGE)
+	{
+		return true;
+	}
+
+	return false;
+}
+
+
+/*
+ * FastShardPruning is a higher level API for FindShardInterval function. Given the relationId
+ * of the distributed table and partitionValue, FastShardPruning function finds the corresponding
+ * shard interval that the partitionValue should be in. FastShardPruning returns NULL if no
+ * ShardIntervals exist for the given partitionValue.
+ */
+static ShardInterval *
+FastShardPruning(Oid distributedTableId, Const *partitionValue)
+{
+	DistTableCacheEntry *cacheEntry = DistributedTableCacheEntry(distributedTableId);
+	int shardCount = cacheEntry->shardIntervalArrayLength;
+	ShardInterval **sortedShardIntervalArray = cacheEntry->sortedShardIntervalArray;
+	bool useBinarySearch = false;
+	uint32 rangeTableId = 1;
+	Var *partitionColumn = PartitionColumn(distributedTableId, rangeTableId);
+	char partitionMethod = PartitionMethod(distributedTableId);
+	FmgrInfo *shardIntervalCompareFunction = NULL;
+	FmgrInfo *hashFunction = NULL;
+	ShardInterval *shardInterval = NULL;
+
+	/* determine whether to use binary search */
+	if (PartitionMethod(distributedTableId) != DISTRIBUTE_BY_HASH ||
```

Comment:
You've already determined the partition method above.

More generally I think it's a bad idea to do a cache lookup for DistributedTableCacheEntry(), PartitionColumn(), PartitionMethod() and PartitionMethod() again. Given that you've looked up the cache entry, just use partitionEntry->partitionMethod instead of calling PartitionMethod() which does a separate hash lookup.